test/jdk/java/net/Socket/CloseOnFailureTest.java

```diff
@@ -0,0 +1,271 @@
+                // Failure will be triggered by the injected `SocketImpl`.
+                new Socket(address, 0xDEAD, address, 0xBEEF);
+            });
+
+            // Run verifications
+            testCase.caughtErrorVerifier.accept(error);
+            testCase.socketImplVerifier.run();
+
+        });
+
+    }
+
+    static List<TestCase> ctor_should_close_on_failures() {
+        return List.of(
+                TestCase.ForBindFailure.ofIOException(),
+                TestCase.ForConnectFailure.ofIOException(),
+                TestCase.ForConnectFailure.ofIllegalArgumentException(1));
+    }
+
+    @Test
+    void connect_should_close_on_unresolved_address() throws IOException {
+        MockSocketImpl socketImpl = new MockSocketImpl(null, null);
+        try (Socket socket = new Socket(socketImpl) {}) {
+            InetSocketAddress address = InetSocketAddress.createUnresolved("no.such.host", 0xBEEF);
+            assertTrue(address.isUnresolved());
+            assertThrows(
+                    UnknownHostException.class,
+                    () -> socket.connect(address, 10_000),
+                    () -> address.getHostName() + " is unresolved");
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("connect_should_close_on_failures")
+    void connect_should_close_on_failures(TestCase testCase) throws Throwable {
+
+        // Create a socket using the mock `SocketImpl` configured to fail
+        try (Socket socket = new Socket(testCase.socketImpl) {}) {
+
+            // Trigger the failure
+            Exception error = assertThrows(Exception.class, () -> {
+                SocketAddress address = Utils.refusingEndpoint();
+                // Address and timeout are mostly ineffective.
+                // They just need to be _valid enough_ to reach to the `SocketImpl#connect()` invocation.
+                // Failure will be triggered by the injected `SocketImpl`.
+                socket.connect(address, 10_000);
+            });
+
+            // Run verifications
+            testCase.caughtErrorVerifier.accept(error);
+            testCase.socketImplVerifier.run();
+            testCase.socketVerifier.accept(socket);
+
+        }
+
+    }
+
+    static List<TestCase> connect_should_close_on_failures() {
+        return List.of(
+                TestCase.ForConnectFailure.ofIOException(),
+                TestCase.ForConnectFailure.ofIllegalArgumentException(0));
+    }
+
+    private record TestCase(
+            String description,
+            MockSocketImpl socketImpl,
+            ThrowingConsumer<Exception> caughtErrorVerifier,
+            ThrowingRunnable socketImplVerifier,
+            ThrowingConsumer<Socket> socketVerifier) {
+
+        private static final String ERROR_MESSAGE = "intentional test failure";
+
+        private static final class ForBindFailure {
```

Comment:
ForConnectionFailure ForBindfFailure are factories abstraction
Their method are factory methods (the world is addicted to of methods) fabricating a TestCase

ofIOException
ofIllegalArgumentException
ofIOException() 

A simple change to ConnectionFailureFactory and BindFailureFactory
With methods like ioExceptionTestCase illegalArgumentException

Lead to readable code like

TestCase.ConnectionFailureFactory.ioExceptionTestCase()

TestCase.BindFaiureFactory.illegalExceptionTestCase()

Reply:
Applied your suggestions in e37ee840b9681aa017f8653df68c69232a6b6c83.